### ISSUE TITLE
Provide a single entrypoint for all public webhooks

### ISSUE THREAD

**Issue by pre** [argoproj/argo-cd#14954]
# Summary

Currently argocd-server acts as a single entrypoint for both the user facing http interface and Dex OIDC requests. Argocd-server forwards OIDC requests to Dex so Dex does not needs it own publicly available endpoint. 

# Motivation

The applicationset-controller webhook currently requires applicationset-controller be publicly available. If applicationset's webhook is routed to argocd-server, argocd-server responds with `Webhook processing failed: event not defined to be parsed` (https://github.com/argoproj/applicationset/issues/516) 

* From service management perspective, it does not seem consistent to expose applicationset-controller's webhook port to the public ingress. 
* Argocd-server already acts as a proxy for the public HTTP requests meant for Dex. Intuitively having the same behaviour for applicationset-controller would be less surprising. 
* The ApplicationSet documentation currently does not mention anything about exposing the applicationset-controller webhook port (https://github.com/argoproj/applicationset/issues/516#issuecomment-1578257277).

Related issue: [Creating a seperate microservice for webhook server in Argo CD and Application Set](https://github.com/argoproj/argo-cd/issues/9378)

# Proposal

In the same way argocd-server forwards OIDC requests to Dex.

**Comment by pre** [comment]
It’s a nice start!

**Comment by tvandinther** [comment]
This makes sense, they do in fact need to be differentiated. I do like the idea of using a URL query parameter to disambiguate the destination as for configuration purposes these are the same endpoint. I could also see a potential scenario where a cluster admin might want to protect the routes differently which would be made more difficult using query parameters. Though I find that the use case of treating them the same and unifying it under one endpoint (and associated configuration) to be more common. Curious to hear other opinions.

**Comment by tvandinther** [comment]
I like this, but perhaps swap the order of the path elements to make wildcard suffixes target both.
 
-  /api/webhook/appset
- /api/webhook/argocd

Then I suppose for backwards compatibility `/api/webhook` operates the same handler as `/api/webhook/argocd`.

**Comment by jmalloc** [comment]
Hi folks, interested onlooker here. I'm not sure I follow the reasoning that the handlers need to be separate.

> So I think just having one single entry point `/api/webhook` for both cases almost impossible, because in that case we don't know which event came.

Isn't the `X-GitHub-Event` header sufficient to tell which kind of event it is? I am unfamiliar with the other systems (GitLab, etc), but it stands to reason that all events must have an equivalent "event type" field, otherwise we couldn't action them in any way.

If the server had a single endpoint, users would no longer need to configure the event producer (GitHub, GitLab, etc) to send the event twice, and therefore it's no longer necessary to determine whether a specific _HTTP request_ is intended for the `appset-controller` or not.

In summary, I'd like to propose:

- Leaving the `/api/webhook` unchanged on all services
- Adding an `/api/webhook/all` endpoint to the server, which:
  - Does the equivalent of `/api/webhook`; and
  - Forwards the event the `appset-controller`; and
  - Potentially forwards to other services in the future

---

To be very explicit about why I think this approach could be beneficial:

As a **user** of Argo CD and GitHub, I don't need to be concerned with which internal component(s) of Argo CD handle each event type, I simply configure GitHub to send `push` and `pull request` events to Argo CD's server.
   
- I no longer need to setup ingress for the `appset-controller`
- I no longer need multiple organisational-level webhook configurations
- I can now use my internal Argo CD GitHub application's webhook configuration
- The payloads are only delivered via the internet once

From a **maintainers** perspective, I see the benefit in an abstraction that allows for new Argo CD services and changes to which event types are handled by which services without reconfiguration on the user's end.

**Comment by tvandinther** [comment]
If a push event is routed to both endpoints, but appset was not intended, all that would happen is the appset controller would drop the event since a generator would presumably not have been configured for that repository.

If an appset does use a generator targeting the repo which already has a webhook configured then it will work without further configuration.

An appset is really the artifact of intention, not the webhook configuration. The webhook only serves to notify Argo CD of an event, something which we are redundantly doing twice.

So while it is true that you can't differentiate the intended target, it could be argued that it doesn't actually matter. Taking the issue back to the core, what we are really asking for is for Argo CD to unify the webhook routing behind a single endpoint. From there it is internal logic which takes over.

Given that core requirement, I think we could explore the option of introducing this on the original `/api/webhook` route. Something to consider in this case is deduplicating events in the appset controller for a version or two while the appset endpoint is deprecated.

**Comment by jmalloc** [comment]
> So while it is true that you can't differentiate the intended target, it could be argued that it doesn't actually matter.

I guess this is what I thought, but didn't articulate exactly!

>  I think we could explore the option of introducing this on the original /api/webhook route.

Even better, even if only because it frees up the most logical endpoint path `/api/webhook` for what I suspect is the behaviour that most users would want.

**Comment by tvandinther** [comment]
The public endpoint would act as a proxy to the internal (potentially microservice) backends. It can accept a connection and keep it open until it receives a response from the real backend (which could be the same HTTP endpoints or a gRPC service if desired).

GitHub will operate with a 10 second timeout to mark a webhook as failed at which point the cancellation can be passed down. I presume other webhook clients do something similar.

In short, you can defer the response until you know the outcome (at least as far as we know based on current implementation).

**Comment by philljie** [comment]
Hi everyone — reviving this thread, and I'd like to align before starting implementation.

I think a single public endpoint(argo-server) serving both webhook types is the right call. Assuming there's agreement on that, the open question is the path format.

Three options based on previous comments in this thread:

  **1. Query parameter** — `/api/webhook?type=appset` / `/api/webhook?type=argocd`

  **2. Path suffix** — `/api/webhook/appset` / `/api/webhook/argocd`
May be useful if a cluster admin wants to protect the routes differently. Not really convinced since webhooks support authorization already.

  **3. Single `/api/webhook`, dispatch server-side** — no change to the URL customers already configure.
Do-able,  but based on my understanding, it is basically asking the argo-server to iterate through all possible backends.  

I am in favor of (1), it fits standard REST resource modeling and minimal changes. @crenshaw-dev , can we have your opinion on this as well?

**Comment by pre** [comment]
Ideally, user should only need to configure a single webhook in a Github repo. With the different paths or parameters, user needs to configure two separate webhooks in the same repository when repo contains both Application and ApplicationSet sources. 

There was thinking above that it’d be difficult to know which webhook should be involved if the path is same for both. Yes, but this needs to be solved anyway because both webhooks may be configured in a single repository. 

-> I’d push for the single endpoint because it’s the cleanest from the user’s perspective, and server needs to differentiate between AppSet and Application webhook in any case.
